drivers/meter/ade7816/ade7816.c

```diff
@@ -0,0 +1,1778 @@
+		irq0_cb.ctx = descriptor;
+		irq0_cb.event = NO_OS_EVT_GPIO;
+		irq0_cb.peripheral = NO_OS_GPIO_IRQ;
+		irq0_cb.handle = init_param->handle_irq0;
+
+		if (init_param->irq0_callback)
+			irq0_cb.callback = init_param->irq0_callback;
+
+		ret = no_os_gpio_get_optional(&descriptor->gpio_irq0_desc,
+					      init_param->gpio_irq0_param);
+		if (ret)
+			goto remove_ade7816;
+
+		ret = no_os_gpio_direction_input(descriptor->gpio_irq0_desc);
+		if (ret)
+			goto remove_irq0;
+
+		ret = no_os_gpio_get_optional(&descriptor->gpio_irq1_desc,
+					      init_param->gpio_irq1_param);
+		if (ret)
+			goto remove_irq0;
+
+		if (descriptor->gpio_irq1_desc) {
+			ret = no_os_gpio_direction_input(descriptor->gpio_irq1_desc);
+			if (ret)
+				goto remove_irq1;
+		}
+
+		ret = no_os_irq_register_callback(init_param->irq_ctrl,
+						  descriptor->gpio_irq0_desc->number, &irq0_cb);
+		if (ret)
+			goto remove_irq1;
+
+		descriptor->irq0_cb = irq0_cb;
+
+		ret = no_os_irq_trigger_level_set(init_param->irq_ctrl,
+						  descriptor->gpio_irq0_desc->number, NO_OS_IRQ_EDGE_FALLING);
+		if (ret)
+			goto remove_irq0_cb;
+
+		ret = no_os_irq_set_priority(init_param->irq_ctrl,
+					     descriptor->gpio_irq0_desc->number, init_param->irq0_priority);
+		if (ret)
+			goto remove_irq0_cb;
+
+		ret = no_os_irq_disable(init_param->irq_ctrl,
+					descriptor->gpio_irq0_desc->number);
+		if (ret)
+			goto remove_irq0_cb;
+
+		break;
+	case ADE7816_IRQ1:
+		if (!init_param->irq_ctrl)
+			goto remove_ade7816;
+
+		irq1_cb.callback = &ade7816_irq1_handler;
+		irq1_cb.ctx = descriptor;
+		irq1_cb.event = NO_OS_EVT_GPIO;
+		irq1_cb.peripheral = NO_OS_GPIO_IRQ;
+		irq0_cb.handle = init_param->handle_irq1;
+
+		if (init_param->irq1_callback)
+			irq1_cb.callback = init_param->irq1_callback;
+
+		ret = no_os_gpio_get_optional(&descriptor->gpio_irq0_desc,
+					      init_param->gpio_irq0_param);
+		if (ret)
+			goto remove_ade7816;
+
+		if (descriptor->gpio_irq0_desc) {
+			ret = no_os_gpio_direction_input(descriptor->gpio_irq0_desc);
+			if (ret)
+				goto remove_irq0;
+		}
+
+		ret = no_os_gpio_get(&descriptor->gpio_irq1_desc,
+				     init_param->gpio_irq1_param);
+		if (ret)
+			goto remove_irq0;
+
+		ret = no_os_gpio_direction_input(descriptor->gpio_irq1_desc);
+		if (ret)
+			goto remove_irq1;
+
+		ret = no_os_irq_register_callback(init_param->irq_ctrl,
+						  descriptor->gpio_irq1_desc->number, &irq1_cb);
+		if (ret)
+			goto remove_irq1;
+
+		descriptor->irq1_cb = irq1_cb;
+
+		ret = no_os_irq_trigger_level_set(init_param->irq_ctrl,
+						  descriptor->gpio_irq1_desc->number, NO_OS_IRQ_EDGE_FALLING);
+		if (ret)
+			goto remove_irq1_cb;
+
+		ret = no_os_irq_set_priority(init_param->irq_ctrl,
+					     descriptor->gpio_irq1_desc->number, init_param->irq1_priority);
+		if (ret)
+			goto remove_irq1_cb;
+
+		ret = no_os_irq_disable(init_param->irq_ctrl,
+					descriptor->gpio_irq1_desc->number);
+		if (ret)
+			goto remove_irq1_cb;
+
+		break;
+	case ADE7816_IRQ0_IRQ1:
```

Comment:
there is a lot of code repetition in these case switches, could you put the repetitive part after the switch and simplify a bit ?